### PR TITLE
feat(scope): support using derived atom in ScopeProvider

### LINKED ISSUE #28: customAtoms are not properly scoped
### Summary

Custom atoms are not scoped because their base atom is not scoped. Is there a workaround for this use-case?

```ts
function customAtom<T>(initialValue) {
  const valueAtom = atom<T>(initialValue)   // <-- valueAtom is not scoped
  return atom(
    (get) => get(valueAtom),
    (_get, set, update: T) => set(valueAtom, update)
  )
}

const someCustomAtom = customAtom(0)

function MyComponent() {
  return (
    <ScopeProvider atoms={[someCustomAtom]}>
      <Counter />
    </ScopeProvider>
  )
}
```

### CodeSandbox Example
https://codesandbox.io/p/sandbox/inspiring-grothendieck-gcc2jt

### Other issues
Not sure if this is related to https://github.com/jotaijs/jotai-scope/issues/24. I don't think it is.

____

## Proposal 💡
https://github.com/jotaijs/jotai-scope/issues/28#issuecomment-2088738873

Scope should apply to dependency trees.

```jsx
const baseAtom = atom('base')
const derivedAtom1 = atom(
  (get) => 'derived1 ' + get(baseAtom)
  (_get, set, value) => {
    // sets the baseAtom scoped to derivedAtom1
    set(baseAtom, value) 
  }
)
const derivedAtom2 = atom((get) => 'derived2 ' + get(baseAtom))

function Component() {
  useHydrate([[derivedAtom1, 'scoped']])

  // deriveAtom1 is scoped, its baseAtom is scoped
  useAtom(derivedAtom1) // 'derived1 scoped'

  // deriveAtom2 is NOT scoped, its baseAtom is NOT scoped
  useAtom(derivedAtom2) // 'derived2 base'

  // baseAtom is NOT directly scoped, baseAtom is NOT scoped
  useAtom(baseAtom)

  ...
}
function App() {
  return (
    <ScopeProvider atoms={[derivedAtom1]} />
      <Component />
    </ScopeProvider> 
  )
}
```

This change will make understanding what is scoped a lot easier.

### Scoping Writable Atoms
```tsx
const baseAtom = atom('base')
const derivedAtom = atom((get) => 'derived ' + get(baseAtom))
const writableAtom = atom(null, (get, set, update) => set(baseAtom, update))

function Component() {
  // deriveAtom1 is scoped, its baseAtom is scoped
  useAtomValue(derivedAtom1)

  // writableAtom is scoped, the baseAtom it writes to is scoped
  // this is the same baseAtom copy used by derivedAtom1
  useSetAtom(writableAtom)

  // baseAtom is NOT directly scoped, baseAtom is NOT scoped
  useAtom(baseAtom)

  ...
}
function App() {
  return (
    <ScopeProvider atoms={[derivedAtom1, writableAtom]} />
      <Component />
    </ScopeProvider> 
  )
}
```

### Revisiting Scoping Base Atoms

```tsx
const baseAtom = atom('base')
const anotherAtom = atom('another')
const derivedAtom = atom((get) => 'derived ' + [get(baseAtom1), get(baseAtom1)])

function Component() {
  // derivedAtom reads the scoped baseAtom and the NOT-scoped anotherAtom
  useAtom(derivedAtom) // 'derived base,another'

  ...
}
function App() {
  return (
    <ScopeProvider atoms={[baseAtom]} />
      <Component />
    </ScopeProvider> 
  )
}
```

### PR DESCRIPTION
**Pull request by yf-yang**
closes #28

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,27 +4,74 @@
 
 https://jotai.org/docs/integrations/scope
 
-## Pro Tips
+## Scoped Atom resolution rules
+1. Each ScopeProvider creates a new scope pool.
+2. If a primitive atom is scoped, its scoped copy will be stored within the scope pool.
+3. If a derived atom is scoped, itself and all of its dependencies will be stored within the scope pool.
```
Comment (dai-shi) [inline README.md:10]: I think it's straightforward and mostly controllable.
I didn't follow the detailed discussion, but I wonder if you have considered this situation.
```js
const a = atom(1);
const b = atom((get) => get(a) * 2);
const c = atom((get) => get(a) * 3);
```
if `b` is explicitly scoped, `a` is implicitly scoped too, and thus `c` is also practically scoped, right?

Comment (yf-yang) [reply README.md:10]: - If b is scoped
a -> global
b -> scoped
c -> global

- if a and b are scoped
a -> scoped
b -> scoped
c -> scoped (implicitly)

File: examples/07_derived/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      <ScopeProvider atoms={[baseAtom]}>
+        <Counter />
+      </ScopeProvider>
+      <h1>Both derived1 an derived2 are scoped</h1>
+      <p>base should be global, derived1 and derived2 are shared</p>
+      <ScopeProvider atoms={[derivedAtom1, derivedAtom2]}>
+        <Counter />
+      </ScopeProvider>
+      <h1>Layer1: Only derived1 is scoped</h1>
+      <p>base and derived2 should be global</p>
```
Comment (yf-yang) [inline examples/07_derived/src/App.tsx:63]: This example here illustrates the behavior

File: README.md (hunk 1 above)
Comment (dai-shi) [reply README.md:10]: Okay, so unless `a` is explicitly scoped, `c` is global. Nice.

**Review by dai-shi — APPROVED** [review]
Although I wasn't able to follow the implementation detail, it looks cool.

File: README.md (hunk 1 above)
Comment (dmaskasky) [reply README.md:10]: Fantastic!

Comment (dmaskasky) [reply README.md:10]: If `b` is scoped, can I write to `b` and have it update implicit scoped `a`?

**Pull request merged**